### PR TITLE
extended `LAMMPS` commands support

### PR DESCRIPTION
**Pull request by cz4rs**
Make it possible to define multiple regions (and the size of simulation box will be adjusted accordingly). This is currently rectricted to `fcc` lattice style only.

Notable changes:
- use `create_atoms` to fill the region with atoms of specific type
- use `mass` to assign mass to atom type
- use `velocity` to assign temperature to atom type

See `in.lb` deck and the screenshots below for practical demonstration.

You can use `paraview --script utils/paraview_process_data.py` to speed up analysis of produced data. It will try to load all `dump*`, `domain_act*` and `domain_lb*` data files from your current directory and do some basic ParaView visualization.

---

TODO (mostly optional):

- [ ] run the simulations for existing decks in CI to ensure no parsing or runtime errors
- [ ] deduce the padding width for filenames from number of steps instead of using hardcoded default
- [ ] improve compatibility with LAMMPS (to be able to run the decks with LAMMPS for comparison)
- [ ] improve `group` command support
- [x] improve communication statistics

---

Multiple regions with different atom types. Note that when the regions overlap, random type will be assigned:
image

Multiple regions with different velocities:
image

Types after 100 steps:
image

Velocities after 100 steps:
image

Types after 1000 steps:
image

### CODE REVIEW

**Comment by cz4rs** [comment]
@streeve This is remaining part of our work on CabanaMD. Marking this as a draft, as there are some limitations (mostly being restricted to `fcc` lattice style only) and in progress bits.
Would you be interested in getting this merged as well?

**Comment by streeve** [comment]
> This is remaining part of our work on CabanaMD. Marking this as a draft, as there are some limitations (mostly being restricted to `fcc` lattice style only) and in progress bits. Would you be interested in getting this merged as well?

Yes definitely, I will take a look

File: paraview_process_data.py
Comment (streeve) [inline paraview_process_data.py:1]: This is definitely useful - maybe we make a `scripts` or `utils` folder?

File: src/cabanamd_impl.h [hunk 1]
```diff
@@ -328,6 +328,7 @@ void CbnMD<t_System, t_Neighbor>::run()
 
             // Exchange atoms across MPI ranks
             comm_timer.reset();
+            log( std::cout, "Comm::exchange()\tstep ", step );
```
Comment (streeve) [inline src/cabanamd_impl.h]: Should this be unified with the rest of the output?

File: src/comm_mpi_impl.h [hunk 2]
```diff
@@ -268,6 +271,11 @@ void Comm<t_System>::exchange()
 
     N_local = N_local + N_total_recv - N_total_send;
 
+    std::cout << '[' << proc_rank << "] "
```
Comment (streeve) [inline src/comm_mpi_impl.h]: Same here - always better to have more output, but I doubt we always want to output this

File: src/inputFile_impl.h [hunk 3]
```diff
@@ -270,12 +273,40 @@ void InputFile<t_System>::check_lammps_command( std::string line,
...
+                     "LAMMPS-Command: 'create_atoms' command only supports "
+                     "'region' option in CabanaMD" );
+        }
     }
     if ( keyword.compare( "mass" ) == 0 )
     {
         known = true;
-        int type = std::stoi( words.at( 1 ) ) - 1;
+        int type = std::stoi( words.at( 1 ) ) -
+                   1; // this is annoying because all other commands have to do
```
Comment (streeve) [inline src/inputFile_impl.h]: I have no problem updating this more broadly if it makes things easier

File: input/in.darma
Comment (streeve) [inline input/in.darma:1]: As someone not very familiar with darma, is it more clear to call this `load_balance` or `lb`?

Comment (cz4rs) [reply input/in.darma:1]: Renamed to `in.lj`.

File: src/inputFile_impl.h (hunk 3 above)
Comment (cz4rs) [reply src/inputFile_impl.h]: Let's leave it as it is, I've removed the comment.

File: paraview_process_data.py
Comment (cz4rs) [reply paraview_process_data.py:1]: Moved to `utils`.

File: src/comm_mpi_impl.h (hunk 2 above)
Comment (cz4rs) [reply src/comm_mpi_impl.h]: Yes, definitely. I will do some testing and polishing to make sure that the information is actually useful.

**Comment by cz4rs** [comment]
@cwschilly This should help with modifying the load balancing test case.

@streeve Thanks for the review. Leaving this as a draft for now, I will fix the output and possibly some other minor issues next week :crossed_fingers:

**Comment by streeve** [comment]
@cz4rs just a note that I updated the formatting to 14 here to match Cabana (#117) and in the near future we'll plan to bump that again to match Kokkos at 16

**Comment by cz4rs** [comment]
> @cz4rs just a note that I updated the formatting to 14 here to match Cabana (#117) and in the near future we'll plan to bump that again to match Kokkos at 16

Alright, I'll rebase this as needed.

File: src/cabanamd_impl.h (hunk 1 above)
Comment (cz4rs) [reply src/cabanamd_impl.h]: Removed (non-significant).

File: src/comm_mpi_impl.h (hunk 2 above)
Comment (cz4rs) [reply src/comm_mpi_impl.h]: Relevant `cabanaMD.out` fragment after integrating this with the other statistics:
```
#Timestep Temperature PotE ETot Time Atomsteps/s LBImbalance Exchanged
0   13.038466   -4.081539   15.472248   0.00    0.00e+00    -nan    -   
10  11.276282   -1.442460   15.468581   0.52    9.67e+04    -nan    -   
20  10.717567   -0.607034   15.466101   1.13    8.11e+04    7.55e-01    1057
30  10.846505   -0.798659   15.467845   1.87    6.77e+04    7.55e-01    -   
40  10.953565   -0.960041   15.467020   2.73    5.80e+04    7.14e-01    1011
50  11.091998   -1.167610   15.467059   3.15    1.19e+05    7.14e-01    -   
60  11.050713   -1.105665   15.467089   3.65    1.02e+05    6.50e-01    704 
70  11.114292   -1.200594   15.467509   4.00    1.41e+05    6.50e-01    -   
80  11.153192   -1.258347   15.468094   4.40    1.26e+05    5.65e-01    639 
90  11.169153   -1.281515   15.468864   4.66    1.89e+05    5.65e-01    -   
100 11.207249   -1.339408   15.468103   4.96    1.66e+05    4.70e-01    555 
```

`Exchanged` column will show up only when Load Balancing is enabled.

Comment (cz4rs) [reply src/comm_mpi_impl.h]: I went with total values since we don't seem to need a breakdown per rank.

**Comment by cz4rs** [comment]
@streeve I have addressed remaining comments. This is also rebased and re-formatted.

File: src/cabanamd_impl.h [hunk 4]
```diff
@@ -681,3 +646,37 @@ void CbnMD<t_System, t_Neighbor>::check_correctness( int step )
...
+                                                 T_V_FLOAT KE, double time,
+                                                 double rate, T_INT exchanged )
+{
+    if ( !_print_lammps )
+    {
+        if ( step == 0 )
+        {
+            log( out, "\n#Timestep Temperature PotE ETot Time Atomsteps/s "
+#ifdef CabanaMD_ENABLE_LB
+                      "LBImbalance Exchanged"
```
Comment (streeve) [inline src/cabanamd_impl.h]:
```suggestion
                      "LB-Imbalance MPI-Exchanged"
```

File: input/in.lb [hunk 5]
```diff
@@ -0,0 +1,41 @@
+# Example using Lennard-Jones potential
```
Comment (streeve) [inline input/in.lb]:
```suggestion
# Example for exercising load balancing (not intended to be physical)
```

**Review by streeve — COMMENTED** [review]
This looks great - couple of wording suggestions, but I think it's ready to merge after that

**Review by streeve — APPROVED** [review]

**Pull request merged**